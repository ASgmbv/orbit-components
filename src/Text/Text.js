// @flow
import * as React from "react";
import defaultTokens from "@kiwicom/orbit-design-tokens";
import styled from "styled-components";
import createComponentFromTagProp from "react-create-component-from-tag-prop";

export const TYPE_OPTIONS = {
  PRIMARY: "primary",
  SECONDARY: "secondary",
  ATTENTION: "attention",
};

export const SIZE_OPTIONS = {
  SMALL: "small",
  NORMAL: "normal",
  LARGE: "large",
};

export const WEIGHT_OPTIONS = {
  NORMAL: "normal",
  BOLD: "bold",
};

export const ALIGN_OPTIONS = {
  LEFT: "left",
  CENTER: "center",
  RIGHT: "right",
};

export const ELEMENT_OPTIONS = {
  P: "p",
  SPAN: "span",
  DIV: "div",
};

type Props = {|
  type: $Values<typeof TYPE_OPTIONS>,
  size: $Values<typeof SIZE_OPTIONS>,
  weight: $Values<typeof WEIGHT_OPTIONS>,
  align: $Values<typeof ALIGN_OPTIONS>,
  italic: boolean,
  uppercase: boolean,
  element: $Values<typeof ELEMENT_OPTIONS>,
  children: React.Node,
  theme: typeof defaultTokens,
|};

const TextElement = createComponentFromTagProp({
  prop: "element",
  propsToOmit: ["tokens", "theme", "type", "size", "weight", "align", "uppercase", "italic"],
});

const StyledText = styled(TextElement)`
  font-family: ${({ theme }) => theme.fontFamily};
  font-size: ${({ tokens, size }) => tokens.sizeText[size]};
  font-weight: ${({ tokens, weight }) => tokens.weightText[weight]};
  color: ${({ tokens, type }) => tokens.colorText[type]};
  line-height: ${({ theme }) => theme.lineHeightText};
  text-align: ${({ align }) => align};
  text-transform: ${({ uppercase }) => uppercase && `uppercase`};
  font-style: ${({ italic }) => italic && `italic`};
  margin: 0;
`;

const Text = (props: Props) => {
  const { children, theme } = props;
  const tokens = {
    colorText: {
      [TYPE_OPTIONS.PRIMARY]: theme.colorTextPrimary,
      [TYPE_OPTIONS.SECONDARY]: theme.colorTextSecondary,
      [TYPE_OPTIONS.ATTENTION]: theme.colorTextAttention,
    },
    weightText: {
      [WEIGHT_OPTIONS.NORMAL]: theme.fontWeightNormal,
      [WEIGHT_OPTIONS.BOLD]: theme.fontWeightBold,
    },
    sizeText: {
      [SIZE_OPTIONS.LARGE]: theme.fontSizeTextLarge,
      [SIZE_OPTIONS.NORMAL]: theme.fontSizeTextNormal,
      [SIZE_OPTIONS.SMALL]: theme.fontSizeTextSmall,
    },
  };
  return (
    <StyledText tokens={tokens} {...props}>
      {children}
    </StyledText>
  );
};

Text.defaultProps = {
  type: "primary",
  size: "normal",
  weight: "normal",
  align: "left",
  element: "p",
  uppercase: false,
  italic: false,
  theme: defaultTokens,
};

export default Text;
