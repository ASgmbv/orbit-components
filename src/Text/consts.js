// @flow
import * as tokens from "@kiwicom/orbit-design-tokens";

export const TYPE_OPTIONS = {
  PRIMARY: "primary",
  SECONDARY: "secondary",
  ATTENTION: "attention",
};

export const SIZE_OPTIONS = {
  SMALL: "small",
  NORMAL: "normal",
  LARGE: "large",
};

export const WEIGHT_OPTIONS = {
  NORMAL: "normal",
  BOLD: "bold",
};

export const ALIGN_OPTIONS = {
  LEFT: "left",
  CENTER: "center",
  RIGHT: "right",
};

export const ELEMENT_OPTIONS = {
  P: "p",
  SPAN: "span",
  DIV: "div",
};

export const colorText = {
  primary: tokens.colorTextPrimary,
  secondary: tokens.colorTextSecondary,
  attention: tokens.colorTextAttention,
};

export const weightText = {
  regular: tokens.fontWeightNormal,
  bold: tokens.fontWeightBold,
};

export const sizeText = {
  large: tokens.fontSizeTextLarge,
  normal: tokens.fontSizeTextNormal,
  small: tokens.fontSizeTextSmall,
};
