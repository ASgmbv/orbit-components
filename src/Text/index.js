// @flow
import * as React from "react";
import * as tokens from "@kiwicom/orbit-design-tokens";

const colorText = {
  primary: tokens.colorTextPrimary,
  secondary: tokens.colorTextSecondary,
  attention: tokens.colorTextAttention,
};

const weightText = {
  regular: tokens.fontWeightNormal,
  bold: tokens.fontWeightBold,
};

const sizeText = {
  large: tokens.fontSizeTextLarge,
  normal: tokens.fontSizeTextNormal,
  small: tokens.fontSizeTextSmall,
};

type Props = {
  type: $Keys<typeof colorText>,
  size: $Keys<typeof sizeText>,
  weight: $Keys<typeof weightText>,
  align: "left" | "center" | "right",
  italic: boolean,
  uppercase: boolean,
  element: "p" | "span" | "div",
  children: React.Node,
  className?: string,
};

function resolveScopedStyles(scope) {
  return {
    className: scope.props.className,
    styles: scope.props.children,
  };
}

const Text = (props: Props) => {
  const { type, size, weight, align, italic, uppercase, element, children, className } = props;

  const scoped = resolveScopedStyles(
    <scope>
      <style jsx>
        {`
          font-family: ${tokens.fontFamily};
          font-size: ${sizeText[size]};
          font-weight: ${weightText[weight]};
          color: ${colorText[type]};
          line-height: ${tokens.lineHeightText};
          text-align: ${align};
          ${uppercase ? `text-transform: uppercase` : ""};
          ${italic ? `font-style: italic` : ""};
          margin: 0;
        `}
      </style>
    </scope>,
  );

  return (
    <React.Fragment>
      {React.createElement(
        element,
        {
          className: [scoped.className, className].join(" "),
        },
        children,
      )}
      {scoped.styles}
    </React.Fragment>
  );
};

Text.defaultProps = {
  type: "primary",
  size: "normal",
  weight: "regular",
  align: "left",
  element: "p",
  uppercase: false,
  italic: false,
};

export default Text;
