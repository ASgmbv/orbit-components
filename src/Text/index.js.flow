// @flow
import type { Props } from "./Text";

declare export default React$ComponentType<Props>;
