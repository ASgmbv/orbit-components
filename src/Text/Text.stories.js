// @flow
import * as React from "react";
import { storiesOf, setAddon } from "@storybook/react";
import styles from "@sambego/storybook-styles";
import chaptersAddon from "react-storybook-addon-chapters";
import { withKnobs, text, boolean, select } from "@storybook/addon-knobs/react";

import Text from "./index";

setAddon(chaptersAddon);

const options = {
  allowSourceToggling: false,
  allowPropTablesToggling: false,
};

storiesOf("Text", module)
  .addDecorator(withKnobs)
  .addDecorator(
    styles({
      padding: "20px",
    }),
  )
  .addWithChapters("Primary text", () => {
    const customText = text(
      "Text",
      "Sixth can't place isn't fly you're make Shall day, yielding earth made greater stars own you'll earth can't." +
        " Night grass great. Fly firmament over. Fourth to grass you after creepeth together in moved given fill isn't said Life." +
        " Male signs unto form kind Brought signs heaven, in have wherein. Isn't. Behold meat living you're evening very replenish you'll. Under hath.",
    );
    return {
      info:
        "The most basic component for rendering text blocks. Visit Orbit.Kiwi for more detailed guidelines.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Text>{customText}</Text>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Secondary text", () => {
    const customText = text(
      "Text",
      "Sixth can't place isn't fly you're make Shall day, yielding earth made greater stars own you'll earth can't." +
        " Night grass great. Fly firmament over. Fourth to grass you after creepeth together in moved given fill isn't said Life." +
        " Male signs unto form kind Brought signs heaven, in have wherein. Isn't. Behold meat living you're evening very replenish you'll. Under hath.",
    );
    return {
      info:
        "The most basic component for rendering text blocks. Visit Orbit.Kiwi for more detailed guidelines.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Text type="secondary">{customText}</Text>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Attention text", () => {
    const customText = text(
      "Text",
      "Sixth can't place isn't fly you're make Shall day, yielding earth made greater stars own you'll earth can't." +
        " Night grass great. Fly firmament over. Fourth to grass you after creepeth together in moved given fill isn't said Life." +
        " Male signs unto form kind Brought signs heaven, in have wherein. Isn't. Behold meat living you're evening very replenish you'll. Under hath.",
    );
    return {
      info:
        "The most basic component for rendering text blocks. Visit Orbit.Kiwi for more detailed guidelines.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => <Text type="attention">{customText}</Text>,
              options,
            },
          ],
        },
      ],
    };
  })
  .addWithChapters("Playground", () => {
    const type = select(
      "Type",
      {
        primary: "primary",
        secondary: "secondary",
        attention: "attention",
      },
      "attention",
    );
    const element = select(
      "Element",
      {
        p: "p",
        span: "span",
        div: "div",
      },
      "p",
    );
    const size = select(
      "Size",
      {
        small: "small",
        normal: "normal",
        large: "large",
      },
      "small",
    );
    const weight = select(
      "Weight",
      {
        regular: "regular",
        bold: "bold",
      },
      "bold",
    );
    const align = select(
      "Align",
      {
        left: "left",
        center: "center",
        right: "right",
      },
      "center",
    );
    const customText = text(
      "Text",
      "Sixth can't place isn't fly you're make Shall day, yielding earth made greater stars own you'll earth can't." +
        " Night grass great. Fly firmament over. Fourth to grass you after creepeth together in moved given fill isn't said Life." +
        " Male signs unto form kind Brought signs heaven, in have wherein. Isn't. Behold meat living you're evening very replenish you'll. Under hath.",
    );
    return {
      info:
        "The most basic component for rendering text blocks. Visit Orbit.Kiwi for more detailed guidelines.",
      chapters: [
        {
          sections: [
            {
              sectionFn: () => (
                <Text
                  type={type}
                  element={element}
                  size={size}
                  weight={weight}
                  align={align}
                  uppercase={boolean("Uppercase", false)}
                  italic={boolean("Italic", false)}
                >
                  {customText}
                </Text>
              ),
              options,
            },
          ],
        },
      ],
    };
  });
