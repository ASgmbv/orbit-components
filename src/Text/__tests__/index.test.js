// @flow
import * as React from "react";
import { shallow } from "enzyme";

import Text from "../index";

describe("Text", () => {
  const text = "Children text";
  const component = shallow(
    <Text type="primary" size="small">
      {text}
    </Text>,
  );
  it("should contain children", () => {
    expect(component.children().text()).toBe(text);
  });
  it("should match snapshot", () => {
    expect(component).toMatchSnapshot();
  });
});
