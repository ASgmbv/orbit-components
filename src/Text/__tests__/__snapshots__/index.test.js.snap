// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendered with children Should contain text 1`] = `
<React.Fragment>
  <p
    className="jsx-1052758011 "
  >
    Children text
  </p>
  <JSXStyle
    css=".__jsx-style-dynamic-selector{font-family:\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif;font-size:12px;font-weight:400;color:#46515e;line-height:1.4;text-align:left;;;margin:0;}"
    dynamic={
      Array [
        "\\"Roboto\\", -apple-system, \\".SFNSText-Regular\\", \\"San Francisco\\", \\"Segoe UI\\", \\"Helvetica Neue\\", \\"Lucida Grande\\", sans-serif",
        "12px",
        "400",
        "#46515e",
        "1.4",
        "left",
        "",
        "",
      ]
    }
    styleId="4206593380"
  />
</React.Fragment>
`;
